src/clients/HubPoolClient.ts

```
@@ -118,20 +120,26 @@ export class HubPoolClient {
   }
 
   // This should find the ProposeRootBundle event whose bundle block number for `chain` is closest to the `block`
-  // without being smaller. It returns the bundle block number for the chain.
+  // without being smaller. It returns the bundle block number for the chain or undefined if not matched.
   getRootBundleEvalBlockNumberContainingBlock(block: number, chain: number, chainIdList: number[]): number | undefined {
     let endingBlockNumber: number;
     for (const rootBundle of sortEventsAscending(this.proposedRootBundles)) {
-      const bundleEvalBlockNumber = this.getBundleEndBlockForChain(
-        rootBundle as ProposedRootBundle,
-        chain,
-        chainIdList
-      );
-      if (bundleEvalBlockNumber >= block) {
-        endingBlockNumber = bundleEvalBlockNumber;
-        // Since events are sorted from oldest to newest, and bundle block ranges should only increase, exit as soon
-        // as we find the first block range that contains the target block.
-        break;
+      try {
+        // `getBundleEndBlockForChain` will error or return the bundle end block for the chain.
+        const bundleEvalBlockNumber = this.getBundleEndBlockForChain(
```

Needed to try-catch this because `getBundleEndBlockForChain` throws if a ProposeRootBundle event isn't the right length. This means that this whole function would ALWAYS throw if the latest proposal was invalid, and then either cancelled or disputed. Instead, we should SKIP these invalid proposals and go to the next most recent one and see if it has a bundle end block that meets the function's search criteria.

This is OK to do because the `dataworker` will check if this function returns undefined.

---

Hm so which errors in `getBundleEndBlockForChain` do you think we should actually throw on? I think they're both ok to skip in this case.

---

The complexity here with my architecture is that `getBundleEndBlockForChain` is also used in `getNextBundleStartBlockNumber` in which case we actually do want it to error. Do you think it'd be better to check some of the assertions from `getBundleEndBlockForChain` before entering the function?

Also I just realized that `getBundleEndBlockForChain` will error if the inputted chainId list length is diff from the inputted rootbundle's block eval list, which means it will error whenever we add a chain to the list.

---

because of that last case, I think its best that `getBundleEndBlockForChain` DOESN'T perform list length checks, only performs the `indexOf` test, and returns 0 if `indexOf == -1`

---

Ok here's a commit that I think simplifies this logic a bit by removing errors thrown https://github.com/across-protocol/relayer-v2/commit/75c1839061fca685e4edac47ed65d1dbb31c4e3a

---

What I'm really trying to get at is that there could be errors that are thrown that are _not directly thrown by your code_. For instance, there could be a bug where you try to call `someMap["abc"].someField` where `someMap["abc"]` is undefined in some unexpected case, causing the whole thing to error.

It seems like you're intending to catch errors that happen in _specific cases_. So for each of those cases, you can throw a specific type of error that you can check for so you don't swallow errors that are a result of bugs and not the intended cases you're trying to capture.

In general, I think we should avoid using try-catch to branch, but sometimes that's unavoidable, so when we do use it, I think it's best to single out the errors you're trying to capture rather than catching everything.

Does that make sense?